### PR TITLE
Workspace with Husky for Simulation and Real Robot Control

### PR DESCRIPTION
**Pull request by YuZhong-Chen**
Closes: #3 #17

### CODE REVIEW

**Review by j3soon — CHANGES_REQUESTED** [review]
Thank you so much for this patch!

Aside from the individual comments, please:
- Separate copying/moving/re-formatting/fixes into different commits.
  - When copying (all or a subset of) files from another repo, be explicit about the commit ids by adding permalinks (i.e., including the commit id in the links). These commits should only include a direct copy, and should not be mixed with any re-formatting or modifications. Please amend the following commits to make the link explicit and separate modifications:
    > To amend these commits, you may want to use interactive rebase (e.g., `git rebase -i HEAD~15`). See this post if you are not familiar with the commands: https://stackoverflow.com/q/1186535
    - 6a319937977ebc62fc519b05c9c67bcf2896d8cb: https://github.com/husky/husky/tree/1e0b1d14d657f04ec3a86e73d6676a2cf7af6f79
    - 2e4060b0de963e5114cb0b9bac2800c23368b528: https://github.com/clearpathrobotics/LMS1xx/tree/90001ac2849c9f6ab9ca9bbfeb816f9b66f966de (I'm not sure about this permalink, since the file contents are different...)
    - 5dbbfe5d6ccdcd40a7c83c4482011e8934b0b1b2: https://github.com/ros-planning/navigation2/blob/2de3f92c0f476de4bda21d1fc5268657b499b258/nav2_bringup/bringup/params/nav2_params.yaml (I'm not sure about this permalink, since the file contents are different...)
  - When moving and modifying/re-formatting many files, it is preferred to isolate the moves/re-format to a dedicated commit, and store the modifications in a separate commit. Otherwise, moving and modifying the files simultaneously may break file history. (See: https://github.com/j3soon/ros2-agv-essentials/pull/10/commits/7a69eaf1e1890131dec7f9569de9ea105625cbc4#diff-61ed84a60fe1f3a3ca688aa7c509713f5a5addca51d4ab1245af4fae904ec893 ) It may be OK to move a few files and make small changes simultaneously in a single commit, as long as the changes are small and obvious. If there's even the slightest chance of causing confusion, simply separate the changes into different commits.
  - As for fixes, it is strongly recommended to keep each commit as simple as possible, even if the fix only requires a few line of changes. It would be even better to include detailed description regarding the corresponding error message, related references for the fix, and the rationale of the fix. See https://github.com/j3soon/ros2-agv-essentials/commit/9da8ce9cb8d13f05756b43b2a57066b3a42a55bf and https://github.com/j3soon/ros2-agv-essentials/commit/12f688907a6967cc2aa52ea80779129b34d957cb for some examples. (Feel free to use more commits, as adding more commits will not result in any additional charges :) )
  - The fundamental principle here is to carefully inspect all changes (potentially with VSCode diff viewer) before adding the file, and include as much relevant information as possible before committing (See: https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=6e96d8a9702d1fef3eb6de47bb84bdea578eb1bb : 20+ lines of description for a one-line change is totally acceptable, and even preferred if all provided information are relevant). The included information should be sufficient for an outsider to understand your modifications just by viewing the commit description and changes. It is also recommended to clean up the commit history before opening a PR (potentially by performing an interactive rebase).
- Minor: Change the directory name from `husky-ws` to `husky_ws` (following the style of `template_ws`).

There are issues in several commits, and should be amended accordingly:
- Commit 2e4060b0de963e5114cb0b9bac2800c23368b528: Please isolate the re-formatting / modifications into another commit.
  - `husky-ws/src/husky/LMS1xx/urdf/sick_lms1xx.urdf.xacro` is said to be directly copied from https://github.com/clearpathrobotics/LMS1xx/blob/humble-devel/urdf/sick_lms1xx.urdf.xacro . However the contents differ.
  - `husky-ws/src/husky/LMS1xx/package.xml` is said to be directly copied from https://github.com/clearpathrobotics/LMS1xx/blob/humble-devel/package.xml . However there are differences in the indentation/whitespaces.
  - `husky-ws/src/husky/LMS1xx/CMakeLists.txt` is said to be directly copied from https://github.com/clearpathrobotics/LMS1xx/blob/humble-devel/CMakeLists.txt . However the contents differ.
- Commit dbcf1a55a997cd550420d704bb4ba79a13b4e778: Please separate the simple re-formatting (quotes/double-quotes, the use of `M_PI`) into another commit, and leave the actual (minimal) fixes in the current commit.
  - There are many changes made in this commit as seen in https://github.com/j3soon/ros2-agv-essentials/pull/10/commits/dbcf1a55a997cd550420d704bb4ba79a13b4e778 . However, there is almost no description on why these changes are necessary, and what issue do they fix. These changes certainly cannot be understood by an outsider, and is definitely not debuggable.
  - If possible, separate the fixes into individual commits based on the error message / issue they fix, and provide description on the rationale of these changes. It is totally acceptable if these fixes need to be separated into 20+ commits.

If you have any additional descriptions or references, please also include them in the commit history or file comments. Please also remove any redundant changes to prevent potential bugs. I would greatly appreciate it if this patch could include as much detail as possible.

<hr>

The comments of this review may seem a bit too obsessive about maintaining a clean commit history. However, please note that this patch is based on many third-party code. It is always beneficial to take extra caution when copying/modifying third-party code. We need to maintain the ability to patch/debug these third-party code in the future. Consider the case where there seems to be a bug in the third-party code sometime in the future:
- If we want to check whether the third-party has provided corresponding patches, we will need the latest commit id of the third-party code included in this repo to quickly determine which patches haven't been included.
- If we want to apply the latest patches, we do not want to deal with broken file history or large changes. We would prefer minimal changes on third-party code.
- If we are unsure about the cause of the bug, detailed commit description for each modifications can help us identify whether the bug is caused by our changes, or is due to the third-party code.
- Etc.

In short, investing extra time in commit time can result in substantial time savings in the future. If there are any undocumented changes in this patch, there is a high chance that we will lose the ability to debug these third-party code and end up reverting this patch and re-implement it again in the future.

<hr>

The comments in this review corresponds to the 7 commits within commit 352c7140340ff48cfc8a890ece671e1ebc5c736e and commit dbcf1a55a997cd550420d704bb4ba79a13b4e778 . I haven't reviewed the rest of the commits yet, but I believe they may contain related issues and should be fixed accordingly before re-requesting review.

Please tidy up the commit history, apply a force-push, and re-request review. Thank you!

File: husky-ws/docker/docker-compose.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      - /tmp/.X11-unix:/tmp/.X11-unix
+      # Direct Rendering Infrastructure
+      - /dev/dri:/dev/dri
+      # Mount sound card to prevent Gazebo warning.
+      - /dev/snd:/dev/snd
+      # Mount Gazebo models directory to reuse models downloaded during first launch.
+      # Reference: https://answers.ros.org/question/365658
+      - ./cache/.gazebo:/home/user/.gazebo
+      # Mount workspace
+      - ../..:/home/ros2-agv-essentials
```
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky-ws/docker/docker-compose.yaml:46]: The `.git` folder is under the `ros2-agv-essentials` folder, if we only mount the working directory into the container, which is `husky_ws`, we cannot access git inside the container.  
  
I think it will be better to mount the outer directory. (It is not necessary, however, it will be convenient)

[the remark this answers is not in the document]
Comment (j3soon) [reply husky-ws/docker/docker-compose.yaml:46]: No problem. I was wondering if we should also update the template workspace to mount the outer directory? (in another PR of course)

[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky-ws/docker/docker-compose.yaml:46]: I will open a PR to fix it later.  
Also, I found a mistake in template_ws, I will fix it together.   

https://github.com/j3soon/ros2-agv-essentials/blob/dba4082f5be7422ad8dd8f4ae3ef8f7544b4db0c/template_ws/.gitignore#L1

Here should be `docker/cache`.

**Comment by ClassLongJoe1112** [comment]
Hi, I had made a world file in order to test the ability for Husky to do U-turn at the corner of a L-shape lane using the default navigation algorithm in `Nav2`. Husky did it well, but I also encountered some problem:

1. Running on my laptop, the FPS in `Gazebo` simulation drops to **13-14** even though I have removed some model objects (like the textured floor) from the original `clearpath_playpen.world` world. I would like to know whether this is because the lack of computing power of my laptop or any other reasons. (In the original world, similar FPS is also observed)
2. I also found the global costmap kind of messy (see [global_costmap_observation](https://github.com/j3soon/ros2-agv-essentials/assets/92087022/885df1a3-a6d9-4499-b106-d0d096530706)), which affects the trajectory planning and I don't know if it only happens to me or not.

This is the world file I tested on:
[clearpath_playpen_corner_test.world.zip](https://github.com/j3soon/ros2-agv-essentials/files/13798060/clearpath_playpen_corner_test.world.zip)

File: husky_ws/README.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+<div align="center">
+    <a href="./">
+        <img src="./figure/view_model.png" width="80%"/>
+    </a>
+</div>
+
+### Demonstration of SLAM.
+
+```bash
+ros2 launch husky_navigation slam_launch.py
```
Comment (j3soon) [inline husky_ws/README.md:82]: The simulation still seems to glitch, especially when the husky robot is rotating. I'm not sure if it's due to the error messages:

```
[planner_server-12] [INFO] [1710316994.361752682] [global_costmap.global_costmap]: Message Filter dropping message: frame 'base_laser
' at time 1524.710 for reason 'the timestamp on the message is earlier than all the data in the transform cache'                     
[controller_server-10] [ERROR] [1710316994.370502888] [tf_help]: Transform data too old when converting from odom to map             
[controller_server-10] [ERROR] [1710316994.370523152] [tf_help]: Data time: 1524s 980000000ns, Transform time: 1524s 710000000ns     
[controller_server-10] [WARN] [1710316994.370544657] [controller_server]: Unable to transform robot pose into global plan's frame    [planner_server-12] [INFO] [1710316994.576282096] [global_costmap.global_costmap]: Message Filter dropping message: frame 'base_laser
' at time 1524.910 for reason 'the timestamp on the message is earlier than all the data in the transform cache'   
```

Do you have any thoughts on the potential fixes for this?

**Review by j3soon — COMMENTED** [review]
Hi, I've finished reviewing this PR. No changes is required.

However, I have a few additional questions and would appreciate it if you could elaborate more on them. Please ignore those comments marked with `Self note`.

I've only tested the simulation part. Will test the real robot control part tomorrow.

File: husky_ws/src/husky/husky_description/launch/description_launch.py [hunk 3]
```diff
@@ -1,13 +1,16 @@
...
-
-
+# This launch file will only launch robot state publisher.
+# It is used when other repos want to get robot state for Husky. For instance: husky_gazebo
 def generate_launch_description():
+    
+    # The path of Husky's velocity controller config.
+    config_husky_velocity_controller = PathJoinSubstitution(
+        [FindPackageShare("husky_control"), "config", "control.yaml"]
+    )
```
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_description/launch/description_launch.py:13]: These lines get the Husky settings that are utilized [here](https://github.com/YuZhong-Chen/ros2-agv-essentials/blob/cd2064fb7485596ddd5b983a4d822d97f552e96b/husky_ws/src/husky/husky_description/urdf/husky.urdf.xacro#L8). 
It's okay if you don't include this parameter in the URDF file, as the default value serves the same purpose.

File: husky_ws/src/husky/husky_gazebo/launch/gazebo.launch.py [hunk 4]
```diff
@@ -20,37 +20,23 @@ def generate_launch_description():
 
     gz_resource_path = SetEnvironmentVariable(name="GAZEBO_MODEL_PATH", value=[
                                                 EnvironmentVariable("GAZEBO_MODEL_PATH", default_value=""),
-                                                "/usr/share/gazebo-11/models/:",
-                                                str(Path(get_package_share_directory("husky_description")).parent.resolve())])
+                                                "/usr/share/gazebo-11/models/",
+                                                ":",
+                                                str(Path(get_package_share_directory("husky_description")).parent.resolve()),
+                                                ":",
+                                                str(Path(get_package_share_directory("realsense2_description")).parent.resolve())])
```
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_gazebo/launch/gazebo.launch.py:27]: This line isn't necessary when using ZED in Gazebo. 
I've included it because the original URDF setting used Realsense as input, and I prefer to keep Realsense from the URDF. It may be more convenient to add ZED in the future.

[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_gazebo/launch/gazebo.launch.py:27]: Therefore, in the ZED to Husky pipeline, we can remove it!

File: husky_ws/docker/Dockerfile [hunk 5]
```diff
@@ -1,5 +1,22 @@
...
+FROM osrf/ros:humble-desktop-full AS amd
+
+# Install Gazebo
+RUN apt-get update && apt-get install -y \
+    ros-$ROS_DISTRO-gazebo-ros-pkgs \
+    && rm -rf /var/lib/apt/lists/*
+
+##########################################################################################
+
+FROM $TARGET_PLATFORM
```
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/docker/Dockerfile:19]: If you want to build the image with `buildx`, you may refer to [here](https://github.com/YuZhong-Chen/ArduinoIDE/blob/7d16d17d45e56e9fa343564d115f7bfb19d8e6e9/.github/workflows/action.yaml#L34-L54).

I believe that using `buildx` is preferable to `docker manifest`, as the `docker manifest` command is experimental and may change between Docker releases without warning, or it could potentially be removed entirely in a future release. [Ref](https://docs.docker.com/reference/cli/docker/manifest/)

File: husky_ws/src/husky/husky_navigation/config/nav2_params.yaml
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:1]: I think I simply copied the style from [here](https://github.com/ros-planning/navigation2_tutorials/blob/24ee184246857328c3622b09e31600d6d5bea754/sam_bot_description/config/nav2_params.yaml#L189) and naively changed the values. 
To be honest, I've forgotten where I obtained these values ...

However, the order of these values won't affect the functionality. We just need to ensure their shape. [Ref](https://navigation.ros.org/setup_guides/footprint/setup_footprint.html)
It's okay to change them to the official ones.

File: husky_ws/README.md (hunk 2 above)
Comment (YuZhong-Chen) [reply husky_ws/README.md:82]: I'm still not sure how to fix this. 
I believe it may be due to some parameters that I haven't set correctly. I'll try to figure it out diligently this week.

File: husky_ws/docker/Dockerfile (hunk 5 above)
[the remark this answers is not in the document]
Comment (j3soon) [reply husky_ws/docker/Dockerfile:19]: Thanks! I believe the commands fit my needs. I'll test them out when configuring the automated build process.

File: husky_ws/src/husky/husky_navigation/config/nav2_params.yaml
[the remark this answers is not in the document]
Comment (j3soon) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:1]: Thanks for the references. I think it's better to stick to the official values to prevent future confusion. Just pushed a commit to fix this.

**Review by j3soon — COMMENTED** [review]
Thanks for fixing the aforementioned simulation issue. The bug seems pretty tricky to catch.

I can confirm the simulation works well on my machine. I'll test the real-world control part soon.

File: husky_ws/src/husky/husky_navigation/config/nav2_params.yaml [hunk 6]
```diff
@@ -125,6 +125,7 @@ controller_server:
     progress_checker_plugin: "progress_checker"
     goal_checker_plugins: ["general_goal_checker"] # "precise_goal_checker"
     controller_plugins: ["FollowPath"]
+    odom_topic: /odometry/filtered
```
[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:128]: I'm uncertain if there's documents available for ROS2, but the document for ROS1 is quite comprehensive. You can refer to this [website](https://docs.ros.org/en/melodic/api/robot_localization/html/state_estimation_nodes.html) for more information.

The `ekf_node` is an implementation of the [Extended Kalman Filter](https://en.wikipedia.org/wiki/Extended_Kalman_filter), which is highly beneficial for sensor fusion tasks. However, comprehending its workings completely can be challenging. While I once had a comprehensive understanding of it, I find that I've forgotten some details over time.

[the remark this answers is not in the document]
Comment (j3soon) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:128]: Thanks for the information. I seem to overlooked the topic name documented in its ROS 1 documentation: https://docs.ros.org/en/noetic/api/robot_localization/html/state_estimation_nodes.html#published-topics

The ROS 2 documentation for robot_localization doesn't seem to exist yet, since Nav2 also refers to the ROS 1 docs: https://navigation.ros.org/setup_guides/odom/setup_odom.html#robot-localization-demo

## Local/Global costmap misalignment

I'm asking this question because I suspect that the `ekf_node` might not be performing sensor fusion as intended. This hypothesis is due to the local costmap and global costmap misalignment mentioned previously. (Your recent fix mitigated the issue, but it still happen sometimes).

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/6dd46462-1bbd-44cb-bd41-5a1d15f5d014

I've used the `rqt_graph` and `ros2 topic` tool and didn't see `ekf_node` using the IMU data.

image

image

This is due to incorrect gazebo IMU plugin topic name. Details of this fix is provided in the first commit below.

## Husky model and odom mismatch

After making `ekf_node` fuse IMU data, the odom pose will diverge from husky model pose.

image

This can be easily fixed by modifying the RVIZ file in the second commit below.

## Large Y error due to IMU fusion

As simulation time passes, the Y error becomes larger and larger.

image

This can be fixed by not using Y data from the IMU as described in the third commit below.

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/d2611c52-026a-4c32-8187-e9e21a068953

## Variance explosion

If the simulation runs for a longer time, the variance of odom estimation still explodes due to not having absolute data.

image

This can be fixed by fusing pose data provided by the slam_toolbox as mentioned in the fourth commit below. However, this causes jumps in odom estimation. I haven't spent time on investigating ways to mitigate this yet.

https://github.com/j3soon/ros2-agv-essentials/assets/20457146/f5df81a4-94e8-4728-9102-c6185ae2b92f

[the remark this answers is not in the document]
Comment (YuZhong-Chen) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:128]: Firstly, I want to express my gratitude for the fixes, especially for the commit addressing the topic issue. I believe it resolves an issue that has been bothering me for quite some time. Specifically, while addressing the previous issue, I attempted to mitigate the localization error by reducing the rotational speed. However, I noticed that the velocity smoother wasn't functioning correctly. At that time, I postponed addressing this issue to the future, but unfortunately, I forgot to resolve it in the end.

---

For directly fusing the pose from `slam_toolbox` into the EKF, I believe it's not a recommended practice due to the jumps in the data, as you mentioned. There is an potential solution:

Enable the config `dynamic_process_noise_covariance` to `true` in the localization.yaml file. This adjustment should help mitigate the issue by allowing the system to adapt its noise covariance dynamically. Remember to disable the fusion of pose in the ekf_node when you want to try this solution. [Reference](https://docs.ros.org/en/melodic/api/robot_localization/html/state_estimation_nodes.html#dynamic-process-noise-covariance)

---

Just a quick note. We should only integrate the global data into the ekf_node, as described below:

Add an additional EKF node to manage the fusion of pose data. Here's how it works: utilize a local EKF node to fuse odometry and IMU data, then employ a global EKF node to fuse the data from the local EKF node and the pose. This approach promises a more robust fusion strategy. You can find the concept explained [here](https://docs.ros.org/en/melodic/api/robot_localization/html/integrating_gps.html#using-navsat-transform-node). However, this strategy may not be applicable to solve the issue here, as we're relying on slam_toolbox to publish the tf map -> odom.

[the remark this answers is not in the document]
Comment (j3soon) [reply husky_ws/src/husky/husky_navigation/config/nav2_params.yaml:128]: Thanks for the information!

Seems like the correct way to go is removing the global pose info and use the `dynamic_process_noise_covariance`. Are you interested in giving it a try? If so, please feel free to make additional commits and remove the global pose information in my last commit.

**Review by j3soon — COMMENTED** [review]
Simulation code works as expected; Real robot control code works well on x86 system, but currently have docker building issues on arm system (Jetson AGX Orin).

* The `docker compose` command requires installing the latest docker & NVIDIA container toolkit on Jetson AGX Orin.

* I recall that the arm system error includes the following when executing `clearpath_computer_installer.sh`:
  ```
  E: Unable to locate package ros-humble-clearpath-robot
  ```

Will continue fixing it tomorrow.

<hr>

2024/05/29 Update:

The issue is due to `ros-humble-clearpath-robot` binaries are only [available on amd64](https://packages.clearpathrobotics.com/stable/ubuntu/dists/jammy/main/binary-amd64/Packages) but [not on arm64](https://packages.clearpathrobotics.com/stable/ubuntu/dists/jammy/main/binary-arm64/Packages).

The straightforward approach is to install package from source when on arm64: https://github.com/clearpathrobotics/clearpath_robot

Will check if this approach works.

**Comment by j3soon** [comment]
Current version can reliably control real husky by either amd64 or arm64 system.

@YuZhong-Chen please help review my commits or perform further changes.

Please re-request a review when no further changes are required, and I'll go ahead and merge this PR. Thanks!

**Comment by YuZhong-Chen** [comment]
Sorry for the late response. 
I have tested on my laptop (amd64) and NV Orin (arm64). I found that on the laptop, the packages like gazebo plugin were not correctly installed inside the container. One reason was a mistake in the command within the 'if' condition in the Dockerfile. Another reason was that the variable `TARGETARCH` was not defined before its use. I have made the necessary corrections, retested on both platforms, and successfully ran them on both sides. Please review this commit again. If everything looks good to you, I think we can merge this PR. Thanks!

**Comment by j3soon** [comment]
Thanks for the fix!

Could you please help check if I missed out any code license we've included in this PR? (see the latest commit)

In addition, should the LiDAR LMS1xx be replaced with VLP-16 LiDAR in this PR? Or is it ok to defer it to another PR some time in the future?

**Comment by YuZhong-Chen** [comment]
I think one license is missing: the license for the udev rule for Husky. See commit [bf09fe1](https://github.com/YuZhong-Chen/ros2-essentials/commit/bf09fe1cddf6d3807e0aeb0eec6dbca138d497cc), which copies the udev rule from [here](https://github.com/clearpathrobotics/clearpath_robot/blob/17d55f1b27d3fe19fb82e7df64dca96dbd345837/clearpath_robot/debian/udev). Since it's a [different repository](https://github.com/clearpathrobotics/clearpath_robot/tree/17d55f1b27d3fe19fb82e7df64dca96dbd345837), we may need to add its license as well.

**Review by j3soon — APPROVED** [review]
Thanks! Looks good to me!
